x/sequencer/keeper/rotation.go

```diff
@@ -95,3 +105,44 @@ func (k Keeper) OnProposerLastBlock(ctx sdk.Context, proposer types.Sequencer) e
 	)
 	return nil
 }
+
+// setSuccessorForRollapp will assign a successor. It won't replace an existing one.
+// It will prioritize non sentinel
+func (k Keeper) setSuccessorForRollapp(ctx sdk.Context, rollapp string) error {
+	successor := k.GetSuccessor(ctx, rollapp)
+	if !successor.Sentinel() {
+		// a valid successor is already set so there's no need to do anything
+		return nil
+	}
+
+	seqs := k.RollappPotentialProposers(ctx, rollapp)
+	successor, err := ProposerChoiceAlgo(seqs)
+	if err != nil {
+		return err
+	}
+	k.SetSuccessor(ctx, rollapp, successor.Address)
+	return nil
+}
+
+// ProposerChoiceAlgo : choose the one with most bond
+// Requires sentinel to be passed in, as last resort.
+func ProposerChoiceAlgo(seqs []types.Sequencer) (types.Sequencer, error) {
+	if len(seqs) == 0 {
+		return types.Sequencer{}, gerrc.ErrInternal.Wrap("seqs must at least include sentinel")
+	}
+	// slices package is recommended over sort package
+	slices.SortStableFunc(seqs, func(a, b types.Sequencer) int {
```

Comment:
shouldn't we filter out the opted-out one?

Reply:
it's filtered out in `RollappPotentialProposers`